evennia/plugin.py

```diff
@@ -0,0 +1,217 @@
+            of Evennia is not guaranteed. ONLY use this method to alter the settings.
+
+        Args:
+            settings (settings.py): The module reference to settings.py.
+            plugins (dict of name->EvPlugin classes): Allows plugins to know what
+                plugins will be loading.
+
+        Returns:
+            None
+        """
+
+    def at_plugin_setup(self):
+        """
+        First phase of plugin setup at the end of evennia._init()
+        """
+
+    def at_plugin_collaborate(self):
+        """
+        Second phase of plugin setup during evennia._init(). At this point, all plugins
+        are available in the evennia.PLUGINS dictionary.
+
+        Use this method for any plugin setup that involves plugins collaborating with each other.
+        """
+
+    def at_plugin_finalize(self):
+        """
+        Third and final phase of plugin setup during evennia._init(). Any last steps? Do them here.
+        """
+
+    def at_server_start(self):
+        """
+        Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_reload_start(self):
+        """
+       Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+       see at_server_startstop module documentation.
+       """
+
+    def at_server_reload_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+    def at_server_cold_start(self):
+        """
+       Runs in the same order as load_priority, but before the 'mygame' hooks.
+
+       see at_server_startstop module documentation.
+       """
+
+    def at_server_cold_stop(self):
+        """
+        Runs in the reverse order of load_priority, but before the 'mygame' hooks.
+
+        see at_server_startstop module documentation.
+        """
+
+
+class PluginManager:
+    """
+    Basic class used for maintaining an index of Plugins installed.
+    """
+
+    def __init__(self):
+        self.plugins = dict()
+        self.plugins_sorted = list()
+        self.reverse = None
+
+    def add(self, plugin):
+        self.plugins[plugin.name()] = plugin
+        self.plugins_sorted.append(plugin)
+
+    def sort(self):
+        self.plugins_sorted.sort(key=lambda x: x.load_priority())
+        self.reverse = reversed(self.plugins_sorted)
+
+    def __getitem__(self, item):
+        return self.plugins[item]
+
+    def init_settings(self, settings):
+        for plugin in self.plugins_sorted:
+            plugin.init_settings(settings, self.plugins)
+
+    def at_plugin_setup(self):
```

Comment:
All these `at_*` methods are mis-named and are confusing. The `at_*` name is used for overloadable hooks that are _called_ at certain times. This is not what happens here - this handler is not itself meant to be overridden and its methods are _calling_ methods (we trigger these at particular server times). So this should really be named `plugin_setup` or maybe `call_plugin_setup` to be very explicit.